Make CompositeEditCommand::prepareWhitespaceAtPositionForSplit() works at deleting SVG element

Ahmad-S792 opened the pull request:
#### 99962e54b060098839e394bb1153ed7bfcb63925
<pre>
Make CompositeEditCommand::prepareWhitespaceAtPositionForSplit() works at deleting SVG element

Make CompositeEditCommand::prepareWhitespaceAtPositionForSplit() works at deleting SVG element
<a href="https://bugs.webkit.org/show_bug.cgi?id=62500">https://bugs.webkit.org/show_bug.cgi?id=62500</a>

Reviewed by Ryosuke Niwa.

Merge - <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=158002

CompositeEditCommand::prepareWhitespaceAtPositionForSplit() replaces collapsible whitespaces with non-breaking space around split position.
This patch changes it to work when split position is end of text node, e.g. VisualPosition.downstream().deprecatedNode() != VisualPostion.deprecateNode().

* Source/WebCore/editing/CompositeEditCommand.h: Add &quot;replaceCollapsibleWhitespaceWithNonBreakingSpaceIfNeeded&quot; header
* Source/WebCore/editing/CompositeEditCommand.cpp:
(CompositeEditCommand::prepareWhitespaceAtPositionForSplit):
(1) Add variable with &quot;previousVisiblePos&quot; and &quot;visiblePos&quot; return values
(2) Add &quot;replaceCollapsibleWhitespaceWithNonBreakingSpaceIfNeeded&quot; function and logic to work with &quot;SVG&quot; deletion
* LayoutTests/editing/deleting/delete-svg-001.html: Add Test Case
* LayoutTests/editing/deleting/delete-svg-001-expected.txt: Add Test Case Expectation

Canonical link: <a href="https://commits.webkit.org/257691@main">https://commits.webkit.org/257691@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b7fac532bb8277639f665074f740ffd1a21e20c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99678 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8855 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32759 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109041 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169272 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103682 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9503 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86145 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92140 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106949 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105450 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90628 "Passed tests") | [💥 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34078 "An unexpected error occured. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Updated gtk dependencies; Running find-modified-layout-tests; Running run-layout-tests-in-stress-mode; layout-tests (exception)") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21990 "Found 1 new test failure: media/modern-media-controls/placard-support/placard-support-airplay-fullscreen.html (failure)") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2678 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23506 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2623 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45889 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5288 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8760 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42980 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4485 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/CompositeEditCommand.cpp:
```diff
@@ -991,12 +991,19 @@ void CompositeEditCommand::prepareWhitespaceAtPositionForSplit(Position& positio
-    if (deprecatedIsCollapsibleWhitespace(visiblePos.characterAfter()) && is<Text>(*position.deprecatedNode()) && !is<HTMLBRElement>(*position.deprecatedNode()))
-        replaceTextInNodePreservingMarkers(downcast<Text>(*position.deprecatedNode()), position.deprecatedEditingOffset(), 1, nonBreakingSpaceString());
+    Position pos = visiblePosition.deepEquivalent().downstream();
+    if (!pos.containerNode() || !is<Text>(*pos.containerNode()) || is<HTMLBRElement>(*pos.containerNode()))
```
I don't think `is<HTMLBRElement>(*pos.containerNode())` makes much sense.
It's probably meant to do `is<HTMLBRElement>(*pos.deprecatedNode())` instead.

Ahmad-S792 replied:
Updated. Thanks!

rniwa reviewed (approved):
r=me

Ahmad-S792 committed:
Make CompositeEditCommand::prepareWhitespaceAtPositionForSplit() works at deleting SVG element

Make CompositeEditCommand::prepareWhitespaceAtPositionForSplit() works at deleting SVG element
https://bugs.webkit.org/show_bug.cgi?id=62500

Reviewed by Ryosuke Niwa.

Merge - https://src.chromium.org/viewvc/blink?view=revision&revision=158002

CompositeEditCommand::prepareWhitespaceAtPositionForSplit() replaces collapsible whitespaces with non-breaking space around split position.
This patch changes it to work when split position is end of text node, e.g. VisualPosition.downstream().deprecatedNode() != VisualPostion.deprecateNode().

* Source/WebCore/editing/CompositeEditCommand.h: Add "replaceCollapsibleWhitespaceWithNonBreakingSpaceIfNeeded" header
* Source/WebCore/editing/CompositeEditCommand.cpp:
(CompositeEditCommand::prepareWhitespaceAtPositionForSplit):
(1) Add variable with "previousVisiblePos" and "visiblePos" return values
(2) Add "replaceCollapsibleWhitespaceWithNonBreakingSpaceIfNeeded" function and logic to work with "SVG" deletion
* LayoutTests/editing/deleting/delete-svg-001.html: Add Test Case
* LayoutTests/editing/deleting/delete-svg-001-expected.txt: Add Test Case Expectation

Canonical link: https://commits.webkit.org/257691@main
